### PR TITLE
Allow plugin factories to be configured for styx test API server.

### LINKED ISSUE #180: Allow test-api StyxServer Builder to take PluginFactory
## The problem
Currently, `com.hotels.styx.testapi.StyxServer.Builder` can take a `Plugin` through a builder method. However, this requires the plugin to have already been instantiated. This prevents the test-api from being able to test how plugin factory behaviour affects Styx, and also denies the plugin access to the `Environment`.

We need to add another method that allows `PluginFactory` to be set in the builder and encourage its use.

## Acceptance criteria
* New method in `Builder` takes `PluginFactory` argument
* When starting Styx, the plugin factory is executed.

### PR DESCRIPTION
**Pull request by mikkokar**
Fixes #180

### CODE REVIEW

File: support/test-api/src/main/java/com/hotels/styx/testapi/StyxServer.java [hunk 1]
```diff
@@ -142,12 +153,36 @@ public MetricRegistry metrics() {
...
+        };
+    }
+
     /**
      * A builder for constructing instances of {@link StyxServer}.
      */
     public static final class Builder {
         private final Map<String, com.hotels.styx.api.service.BackendService> routes = new HashMap<>();
-        private final List<NamedPlugin> plugins = new ArrayList<>();
+        public final List<PluginTriple> pluginFactories = new ArrayList<>();
```
Comment (kvosper) [inline support/test-api/src/main/java/com/hotels/styx/testapi/StyxServer.java]: This field should be private.

File: support/test-api/src/main/java/com/hotels/styx/testapi/StyxServer.java [hunk 2]
```diff
@@ -208,4 +248,16 @@ public StyxServer start() {
             return new StyxServer(this).start();
         }
     }
+
+    private static class PluginTriple {
```
Comment (kvosper) [inline support/test-api/src/main/java/com/hotels/styx/testapi/StyxServer.java]: The name `Triple` describes the structure of the class, but the name should describe its purpose. The structure is clear from the code.

File: support/test-api/src/main/java/com/hotels/styx/testapi/StyxServer.java [hunk 3]
```diff
@@ -208,4 +248,16 @@ public StyxServer start() {
...
+        private final String name;
+        private final PluginFactory pluginFactory;
+        private final Object pluginConfig;
+
+        PluginTriple(String name, PluginFactory pluginFactory, Object pluginConfig) {
+            this.name = name;
```
Comment (kvosper) [inline support/test-api/src/main/java/com/hotels/styx/testapi/StyxServer.java]: These should be null-checked.

File: support/test-api/src/main/java/com/hotels/styx/testapi/StyxServer.java (hunk 2 above)
Comment (kvosper) [reply support/test-api/src/main/java/com/hotels/styx/testapi/StyxServer.java]: Maybe `PluginConfiguration` or `PluginFactoryConfiguration`.

File: support/test-api/src/main/java/com/hotels/styx/testapi/StyxServer.java [hunk 4]
```diff
@@ -142,12 +153,36 @@ public MetricRegistry metrics() {
         return metricRegistry;
     }
 
+    private PluginFactory.Environment toPluginEnvironment(Environment environment, PluginTriple x) {
```
Comment (kvosper) [inline support/test-api/src/main/java/com/hotels/styx/testapi/StyxServer.java]: `x` is not a good variable name.

Comment (kvosper) [reply support/test-api/src/main/java/com/hotels/styx/testapi/StyxServer.java]: This method can be static.

Comment (mikkokar) [reply support/test-api/src/main/java/com/hotels/styx/testapi/StyxServer.java]: Addressed issues.

File: support/test-api/src/main/java/com/hotels/styx/testapi/StyxServer.java (hunk 3 above)
Comment (mikkokar) [reply support/test-api/src/main/java/com/hotels/styx/testapi/StyxServer.java]: fixed

File: support/test-api/src/main/java/com/hotels/styx/testapi/StyxServer.java (hunk 2 above)
Comment (mikkokar) [reply support/test-api/src/main/java/com/hotels/styx/testapi/StyxServer.java]: fixed

File: support/test-api/src/main/java/com/hotels/styx/testapi/StyxServer.java (hunk 1 above)
Comment (mikkokar) [reply support/test-api/src/main/java/com/hotels/styx/testapi/StyxServer.java]: fixed

**Pull request merged**